numpyro/util.py

```diff
@@ -21,7 +21,7 @@
 from jax.tree_util import tree_flatten, tree_map
 
 _DISABLE_CONTROL_FLOW_PRIM = False
-CHAIN_RE = re.compile(r"\d+$")
+CHAIN_RE = re.compile(r"(?<=_)\d+$")  # e.g. get '3' from 'TFRT_CPU_3'
```

Comment:
Oops, sorry, I meant `_CHAIN_RE`

Reply:
ah, no worries, have updated (I guess it doesn't hurt to keep the underscore in the pattern too?)

The previous CI run had this failure:
```
=================================== FAILURES ===================================
________________________ test_functional_map[pmap-NUTS] ________________________

algo = 'NUTS', map_fn = <function pmap at 0x7fd3df8802f0>

    @pytest.mark.parametrize("algo", ["HMC", "NUTS"])
    @pytest.mark.parametrize("map_fn", [vmap, pmap])
    @pytest.mark.skipif(
        "XLA_FLAGS" not in os.environ,
        reason="without this mark, we have duplicated tests in Travis",
    )
    def test_functional_map(algo, map_fn):
        if map_fn is pmap and xla_bridge.device_count() == 1:
            pytest.skip("pmap test requires device_count greater than 1.")
    
        true_mean, true_std = 1.0, 2.0
        num_warmup, num_samples = 1000, 8000
    
        def potential_fn(z):
            return 0.5 * jnp.sum(((z - true_mean) / true_std) ** 2)
    
        init_kernel, sample_kernel = hmc(potential_fn, algo=algo)
        init_params = jnp.array([0.0, -1.0])
        rng_keys = random.split(random.PRNGKey(0), 2)
    
        init_kernel_map = map_fn(
            lambda init_param, rng_key: init_kernel(
                init_param, trajectory_length=9, num_warmup=num_warmup, rng_key=rng_key
            )
        )
        init_states = init_kernel_map(init_params, rng_keys)
    
        fori_collect_map = map_fn(
            lambda hmc_state: fori_collect(
                0,
                num_samples,
                sample_kernel,
                hmc_state,
                transform=lambda x: x.z,
                progbar=False,
            )
        )
        chain_samples = fori_collect_map(init_states)
    
        assert_allclose(
>           jnp.mean(chain_samples, axis=1), jnp.repeat(true_mean, 2), rtol=0.06
        )
E       AssertionError: 
E       Not equal to tolerance rtol=0.06, atol=0
E       
E       Mismatched elements: 1 / 2 (50%)
E       Max absolute difference: 50.5725
E       Max relative difference: 50.5725
E        x: array([  1.020785, -49.5725  ], dtype=float32)
E        y: array([1., 1.], dtype=float32)

test/infer/test_mcmc.py:752: AssertionError
```

I presume it's unrelated?

Reply:
It passed this time, so I guess so

Reply:
jax program is typically deterministic (unless there are versioning updates). I'll take a closer look when this happens again. Thanks for pointing out!